query/transformation_test.go

```diff
@@ -0,0 +1,282 @@
+				},
+			},
+			{
+				Values: []float64{0, 1, 2, 3, 2, 1},
+				TagSet: api.TagSet{
+					"series": "C",
+				},
+			},
+		},
+		Timerange: api.Timerange{
+			Start:      758300,
+			End:        758300 + 30*5,
+			Resolution: 30,
+		},
+		Name: "test",
+	}
+	testCases := []struct {
+		transform transform
+		parameter []value
+		expected  map[string][]float64
+	}{
+		{
+			transform: transformDerivative,
+			parameter: []value{},
+			expected: map[string][]float64{
+				"A": {0, 1.0 / 30, 1.0 / 30, 1.0 / 30, 1.0 / 30, 1.0 / 30},
+				"B": {0, 0, -1.0 / 30, 0, 2.0 / 30, 0},
+				"C": {0, 1.0 / 30, 1.0 / 30, 1.0 / 30, -1.0 / 30, -1.0 / 30},
+			},
+		},
+		{
+			transform: transformIntegral,
+			parameter: []value{},
+			expected: map[string][]float64{
+				"A": {0, 1 * 30, 3 * 30, 6 * 30, 10 * 30, 15 * 30},
+				"B": {2 * 30, 4 * 30, 5 * 30, 6 * 30, 9 * 30, 12 * 30},
+				"C": {0, 1 * 30, 3 * 30, 6 * 30, 8 * 30, 9 * 30},
+			},
+		},
+		{
+			transform: transformMovingAverage,
+			parameter: []value{scalarValue(100)}, // 100 seconds corresponds to roughly 3 samples
+			expected: map[string][]float64{
+				"A": {0, 0.5, 1, 2, 3, 4},
+				"B": {2.0, 2.0, 5.0 / 3, 4.0 / 3, 5.0 / 3, 7.0 / 3},
+				"C": {0, 0.5, 1, 2, 7.0 / 3, 2},
+			},
+		},
+	}
+	for _, test := range testCases {
+		result, err := ApplyTransform(list, test.transform, test.parameter)
+		if err != nil {
+			t.Error(err)
+			continue
+		}
+		alreadyUsed := make(map[string]bool)
+		for _, series := range result.Series {
+			name := series.TagSet["series"]
+			expected, ok := test.expected[name]
+			if !ok {
+				t.Errorf("Series not present in testcase (A, B, or C). Is instead [%s]", name)
+				continue
+			}
+			if alreadyUsed[name] {
+				t.Errorf("Multiple series posing as %s", name)
+				continue
+			}
+			alreadyUsed[name] = true
+			// Lastly, compare the actual values
+			if len(series.Values) != len(expected) {
+				t.Errorf("Expected result to have %d entries but has %d entries; for series %s", len(expected), len(series.Values), name)
+				continue
+			}
+			// Check that elements are within epsilon
+			for i := range series.Values {
+				if math.Abs(series.Values[i]-expected[i]) > epsilon {
+					t.Errorf("Expected values for series %s to be %+v but are %+v", name, expected, series.Values)
+					break
+				}
+			}
+		}
+	}
+}
+
+func TestApplyTransformFailure(t *testing.T) {
+	list := api.SeriesList{
+		Series: []api.Timeseries{
+			{
+				Values: []float64{0, 1, 2, 3, 4, 5},
+				TagSet: api.TagSet{
+					"series": "A",
+				},
+			},
+			{
+				Values: []float64{2, 2, 1, 1, 3, 3},
+				TagSet: api.TagSet{
+					"series": "B",
+				},
+			},
+			{
+				Values: []float64{0, 1, 2, 3, 2, 1},
+				TagSet: api.TagSet{
+					"series": "C",
+				},
+			},
+		},
+		Timerange: api.Timerange{
+			Start:      758300,
+			End:        758300 + 30*5,
+			Resolution: 30,
+		},
+		Name: "test",
+	}
+	testCases := []struct {
+		transform transform
+		parameter []value
+		expected  string
```

Comment:
This shouldn't compare expected error messages. It's really fragile (tweaking the message slightly means you have to update all these tests). Just testing existence of the error should be sufficient.